tests/test_swift.py

```diff
@@ -87,35 +73,139 @@ def setUp(self):
+
+    def test_channel_names(self):
+        channelNames = self.swift.channelNames()
+        for channelName in self.channels:
+            self.assertIn(channelName, channelNames)
+
+    def test_subscriber_names(self):
+        for channel in self.channels:
+            subscriberNames = self.swift.subscriberNames(channel)
+            for name, info in APP_INFOS.items():
+                if channel in info.channel:
+                    self.assertIn(name, subscriberNames)
+
+    def test_unsubcribe(self):
+        self.assertEqual(self.swift.unsubscribe("app1", "ch1"), True)
+        self.assertNotIn("app1", self.swift._subscribers["ch1"])
+        self.assertEqual(self.swift.unsubscribe("app2", "ch1"), False)
+
+    def test_swiftcall(self):
+        QMessageBox.warning = MagicMock(return_value=QMessageBox.Ok)
+        self.swift._swiftcall("app1", json.dumps({"call": "appNames", "args": {}}))
+        self.swift._swiftcall("app1", json.dumps({
+            "call": "createApp",
+            "args": {"name": "app3", "info": {"module": "module2", "cls": "cls2"}}
+        }))
+        self.swift.callWithSerializable = MagicMock()
+        self.swift.getSerializable = MagicMock(return_value=APP_INFOS["app1"])
+        self.swift._swiftcall("app1", json.dumps({"call": "getSerializable", "args": {}}))
+        self.swift._swiftcall("app1", json.dumps({"call": "_broadcast", "args": {}}))
+        QMessageBox.warning.return_value = QMessageBox.Cancel
+        self.swift._swiftcall("app1", json.dumps({"call": "appNames", "args": {}}))
+
     def test_broadcast(self):
         for channelName in self.channels:
             self.swift._broadcast(channelName, "test_msg")
         for name, app_ in self.swift._apps.items():
             self.assertEqual(len(APP_INFOS[name].channel), app_.received.emit.call_count)
 
 
+class BaseAppTest(unittest.TestCase):
+    """Unit test for BaseApp class."""
+
+    def setUp(self):
+        self.app = swift.BaseApp("name")
+
+    def test_init(self):
+        self.assertEqual(self.app.name, "name")
+        self.assertIsInstance(self.app.swiftcall, swift.SwiftcallProxy)
+        self.assertEqual(type(self.app.broadcastRequested), pyqtBoundSignal)
+        self.assertEqual(type(self.app.received), pyqtBoundSignal)
+        self.assertEqual(type(self.app.swiftcallRequested), pyqtBoundSignal)
+        self.assertEqual(type(self.app.swiftcallReturned), pyqtBoundSignal)
+
+    def test_set_parent(self):
+        swift.BaseApp("name", QObject())
+
+    def test_frames(self):
+        self.assertIsInstance(self.app.frames(), Iterable)
+
+    def test_broadcast(self):
+        self.app.broadcast("ch1", "msg")
+        self.app.broadcast("ch1", lambda: None)
+
+    def test_received_message(self):
+        self.app.receivedSlot = MagicMock()
+        self.app._receivedMessage("ch1", '"msg"')
+        self.app.receivedSlot.assert_called_once()
+        self.app._receivedMessage("ch1", '"msg1" "msg2"')
```

Comment:
What is this code for?

Reply:
Similar to the above question, it is for checking to raise a `JSONDecodeError`.

Reply:
However it does not _check_ whether the `JSONDecodeError` occurred.
Can we know if it worked without exceptions or not?

Reply:
Then maybe checking whether `receivedSlot()` is called would be the only way...